src/utils/waker.rs

```
@@ -50,26 +51,32 @@ impl Readiness {
     }
 }
 
+#[derive(Debug, Clone)]
 pub(crate) struct StreamWaker {
     id: usize,
     readiness: Arc<Mutex<Readiness>>,
-    parent: Waker,
+    parent_waker: Option<Waker>,
 }
 
 impl StreamWaker {
-    pub(crate) fn new(id: usize, readiness: Arc<Mutex<Readiness>>, parent: Waker) -> Self {
+    pub(crate) fn new(id: usize, readiness: Arc<Mutex<Readiness>>) -> Self {
         Self {
             id,
             readiness,
-            parent,
+            parent_waker: None,
         }
     }
+
+    pub(crate) fn set_parent_waker(&mut self, parent: Waker) {
+        self.parent_waker = Some(parent);
+    }
 }
 
 impl Wake for StreamWaker {
     fn wake(self: std::sync::Arc<Self>) {
         if !self.readiness.lock().unwrap().set_ready(self.id) {
-            self.parent.wake_by_ref()
+            let parent = self.parent_waker.as_ref().expect("No parent waker was set");
```

It's probably okay to silently do nothing if the parent waker is not set.